File: ui-automation-framework/pages/login.page.ts
import type { Locator, Page } from '@playwright/test';
import type { Credentials } from '../data/users';

export class LoginPage {
  readonly usernameInput: Locator;
  readonly passwordInput: Locator;
  readonly loginButton: Locator;
  readonly errorMessage: Locator;
  readonly errorCloseButton: Locator;

  constructor(readonly page: Page) {
    this.usernameInput = page.getByTestId('username');
    this.passwordInput = page.getByTestId('password');
    this.loginButton = page.getByTestId('login-button');
    this.errorMessage = page.getByTestId('error');
    this.errorCloseButton = page.getByTestId('error-button');
  }

  async goto(): Promise<void> {
    await this.page.goto('/');
  }

  /** Fills only the non-empty fields so the same method covers negative cases. */
  async login({ username, password }: Credentials): Promise<void> {
    if (username) {
      await this.usernameInput.fill(username);
    }
    if (password) {
      await this.passwordInput.fill(password);
    }
    await this.loginButton.click();
  }
}

File: ui-tests/src/test/java/saucedemo/pages/LoginPage.java
package saucedemo.pages;

import com.microsoft.playwright.Locator;
import com.microsoft.playwright.Page;

public class LoginPage {
    private final Page page;

    public LoginPage(Page page) {
        this.page = page;
    }

    public void open() {
        page.navigate("/");
    }

    public Locator usernameField() {
        return page.locator("[data-test='username']");
    }

    public Locator passwordField() {
        return page.locator("[data-test='password']");
    }

    public Locator loginButton() {
        return page.locator("[data-test='login-button']");
    }

    public Locator error() {
        return page.locator("[data-test='error']");
    }

    public Locator errorClose() {
        return page.locator("[data-test='error-button']");
    }

    public void login(String user, String pass) {
        if (user != null && !user.isEmpty()) {
            usernameField().fill(user);
        }
        if (pass != null && !pass.isEmpty()) {
            passwordField().fill(pass);
        }
        loginButton().click();
    }
}
